### PR TITLE
Implement `Infinite` `Boundary` Conditions for `MatrixProduct` and `ProjectedEntangledPair` Types

### LINKED ISSUE #77: Implement `Infinite` boundary conditions for `MatrixProduct` and `ProjectedEntangledPair`
### Summary
`MatrixProduct` and `ProjectedEntangledPair` currently have `Open` and `Periodic` boundary conditions. We could enhance these structs by adding the `Infinite` boundary condition, in which a cell-like structure is repeated infinitely many times.

### PR DESCRIPTION
**Pull request by jofrevalles**
### Summary

In this PR, we resolve #77 by extending our `MatrixProduct` and `ProjectedEntangledPair` types to include support for infinite boundary conditions, materialized in the infinite Matrix Product States (iMPS) and infinite Projected Entangled Pair States (iPEPS). This creates a structure where the tensors cycle periodically, emulating an infinite system by using a finite system unit cell that repeats itself. 

The `tensors`, `Base.show` and `Base.length` functions have been extended to support `TensorNetworks` with `Infinte` boundary conditions. Additional unit tests have been added to ensure the proper functioning of these new boundary conditions for both the `MPS` and `MPO` structures.

Nevertheless, the visualization of infinite boundary types is not yet supported and we should think about how to implement this.

### Example
Let's create a `MatrixProduct{State, Infinite}` which consists of a cell of three tensors that repeat indefinitely:
```julia
julia> _arrays = [rand(1, 1, 2), rand(1, 1, 2), rand(1, 1, 2)]
3-element Vector{Array{Float64, 3}}:
 [0.17136137178613753;;; 0.1659622523369414]
 [0.23387611601182623;;; 0.5187233455849274]
 [0.025368244393775274;;; 0.2701565528371912]

julia> ψ = MatrixProduct{State,Infinite}(_arrays, order = (:l, :r, :o))
TensorNetwork{MatrixProduct{State, Infinite}, NamedTuple{(:plug, :interlayer, :χ), Tuple{Type{<:Plug}, Vector{Bijection{Int64, Symbol}}, Union{Nothing, Int64}}}}(#tensors=∞, #labels=∞)

julia> tensors(ψ, 1) == tensors(ψ, 4) == tensors(ψ, -2)
true
```

### CODE REVIEW

File: src/Quantum/MP.jl [hunk 1]
```diff
@@ -110,6 +111,17 @@ end
 const MPS = MatrixProduct{State}
 const MPO = MatrixProduct{Operator}
 
+tensors(ψ::TensorNetwork{MatrixProduct{P,Infinite}}, args...) where {P<:Plug} =
+    throw(ArgumentError("You need to specify the site for an infinite MatrixProduct$P"))
+
```
Comment (mofeing) [inline src/Quantum/MP.jl:116]: If it doesn't match any method signature, it's better to let Julia throw a `MethodError`.
```suggestion
```

File: src/Quantum/PEP.jl [hunk 2]
```diff
@@ -116,6 +117,17 @@ end
 const PEPS = ProjectedEntangledPair{State}
 const PEPO = ProjectedEntangledPair{Operator}
 
+tensors(ψ::TensorNetwork{ProjectedEntangledPair{P,Infinite}}, args...) where {P<:Plug} =
+    throw(ArgumentError("You need to specify the site for an infinite MatrixProduct$P"))
+
```
Comment (mofeing) [inline src/Quantum/PEP.jl:122]: Same as in `MP.jl`
```suggestion
```

File: test/MatrixProductOperator_test.jl [hunk 3]
```diff
@@ -76,6 +76,14 @@
...
                 end
             end
+
+            @testset "tensors" begin
+                arrays = [rand(1, 2, 2), rand(1, 1, 2, 2), rand(1, 2, 2)]
+                ψ = MatrixProduct{Operator,Open}(arrays, order = (:l, :r, :i, :o))
+
+                @test tensors(ψ, 1) isa Tensor
+                @test size(tensors(ψ)) |> first == length(ψ) == 3
+            end
```
Comment (mofeing) [inline test/MatrixProductOperator_test.jl:86]: Again you mixed PRs.
```suggestion
```

File: src/Quantum/MP.jl (hunk 1 above)
Comment (jofrevalles) [reply src/Quantum/MP.jl:116]: Sorry, I `cherry-picked` the commits in #78 because I thought was necessary, I will try to remove them now.

File: src/Quantum/MP.jl [hunk 4]
```diff
@@ -110,6 +111,17 @@ end
 const MPS = MatrixProduct{State}
 const MPO = MatrixProduct{Operator}
 
+tensors(ψ::TensorNetwork{MatrixProduct{P,Infinite}}, args...) where {P<:Plug} =
+    throw(MethodError("You need to specify the site for an infinite MatrixProduct$P"))
+
```
Comment (mofeing) [inline src/Quantum/MP.jl:116]: I meant don't even put a method jajaja. Sorry if I didn't explain it well.

The Julia Runtime will already throw a `MethodError` if a matching method doesn't exist.

```suggestion
```

File: src/Quantum/PEP.jl [hunk 5]
```diff
@@ -116,6 +117,17 @@ end
 const PEPS = ProjectedEntangledPair{State}
 const PEPO = ProjectedEntangledPair{Operator}
 
+tensors(ψ::TensorNetwork{ProjectedEntangledPair{P,Infinite}}, args...) where {P<:Plug} =
+    throw(MethodError("You need to specify the site for an infinite MatrixProduct$P"))
+
```
Comment (mofeing) [inline src/Quantum/PEP.jl:122]:
```suggestion
```

File: src/Quantum/MP.jl [hunk 6]
```diff
@@ -110,6 +111,17 @@ end
...
+tensors(ψ::TensorNetwork{MatrixProduct{P,Infinite}}, site::Int, args...) where {P<:Plug} =
+    tensors(plug(ψ), ψ, mod1(site, length(ψ.tensors)), args...)
+
```
Comment (mofeing) [inline src/Quantum/MP.jl:119]: This should already be provided by the `tensors` method in `Quantum.jl`. Why reimplement this?
```suggestion
```

File: src/Quantum/MP.jl [hunk 7]
```diff
@@ -110,6 +111,17 @@ end
...
+Base.show(io::IO, ψ::TensorNetwork{MatrixProduct{P,Infinite}}) where {P<:Plug} =
+    print(io, "$(typeof(ψ))(#tensors=∞, #labels=∞)")
+
```
Comment (mofeing) [inline src/Quantum/MP.jl:122]: I think this method shouldn't be overridden, but just the `length` method specialized. Maybe the number of labels it's difficult to obtain?
```suggestion
```

File: src/Quantum/PEP.jl [hunk 8]
```diff
@@ -116,6 +117,17 @@ end
...
+tensors(ψ::TensorNetwork{ProjectedEntangledPair{P,Infinite}}, site::Int, args...) where {P<:Plug} =
+    tensors(plug(ψ), ψ, mod1(site, length(ψ.tensors)), args...)
+
+Base.show(io::IO, ψ::TensorNetwork{ProjectedEntangledPair{P,Infinite}}) where {P<:Plug} =
+    print(io, "$(typeof(ψ))(#tensors=∞, #labels=∞)")
+
```
Comment (mofeing) [inline src/Quantum/PEP.jl:128]: Same as in MP.jl
```suggestion
```

**Review by mofeing — CHANGES_REQUESTED** [review]

File: src/Quantum/MP.jl (hunk 6 above)
Comment (jofrevalles) [reply src/Quantum/MP.jl:119]: Well, this is for the specific case of an `Infinite` chain of tensors. For example, if we have two tensors that repeat indefinitely, then `tensors(3)` should return `tensors(1)`. This is different from the `Open` and `Periodic` boundary conditions.

File: src/Quantum/MP.jl (hunk 7 above)
Comment (jofrevalles) [reply src/Quantum/MP.jl:122]: Yes, that's it. Since `labels=length(tn.indices)` it was difficult to fix that without overwritting it. However, I think we can change `labels=length(labels(tn))` and this should be fixed.

**Review by mofeing — APPROVED** [review]
